### PR TITLE
[API][Behat] Add note about behat client changes

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.12                   |
| Bug fix?        | no                                                     |
| New feature?    | no                                                     |
| BC breaks?      | no                                                   |
| Deprecations?   | no  |
| License         | MIT                                                          |

Related to [PR](https://github.com/Sylius/Sylius/pull/13832) and [PR](https://github.com/Sylius/Sylius/pull/13869)

### CODE REVIEW

File: UPGRADE-1.12.md [hunk 1]
```diff
@@ -84,6 +84,115 @@ with test or remove these services with complier pass.
...
+  +  $this->client->buildUploadRequest(Resources::AVATAR_IMAGES);
+  ```
+
+As you can see the `Sylius\Behat\Context\Api\Resources` class has all the defined clients stored in constances.
+We also removed the `Sylius\Behat\ClientApiIriClient` alongside with `Sylius\Behat\ClientApiIriClientInterface` as it's no more used.
+
+#### Changes in contexts
+
+The constructors of the behat contexts changed with the improvement.
+Here is the example change of the constructor of `Sylius\Behat\Context\Api\Shop\OrderContexts`
```
Comment (NoResponseMate) [inline UPGRADE-1.12.md]:
```suggestion
Here is the example change of the constructor of `Sylius\Behat\Context\Api\Shop\OrderContext`
```
Though IMO, it'd be better to show an example where the number of clients actually changed.

File: UPGRADE-1.12.md [hunk 2]
```diff
@@ -84,6 +84,115 @@ with test or remove these services with complier pass.
...
+
+With the follow change we are no more defining the clients in `api.xml` file (there are only defined the `shop` and `admin` endpoints) and now to make a call by a different client you have to pass it inside of the methods above.
+You can see the actual difference in example below:
+
+  ```diff
+  -  $this->avatarImagesClient->buildUploadRequest();
+  +  $this->client->buildUploadRequest(Resources::AVATAR_IMAGES);
+  ```
+
+As you can see the `Sylius\Behat\Context\Api\Resources` class has all the defined clients stored in constances.
```
Comment (NoResponseMate) [inline UPGRADE-1.12.md]:
```suggestion
The `Sylius\Behat\Context\Api\Resources` class contains constants for all the defined resources.
```

File: UPGRADE-1.12.md [hunk 3]
```diff
@@ -84,6 +84,115 @@ with test or remove these services with complier pass.
...
+  +  public function buildCreateRequest(string $resource): void;
+  
+  -  public function buildUpdateRequest(string $id): void;
+  +  public function buildUpdateRequest(string $resource, string $id): void;
+  
+  -  public function buildUploadRequest(): void;
+  +  public function buildUploadRequest(string $resource): void;
+  ```
+
+With the follow change we are no more defining the clients in `api.xml` file (there are only defined the `shop` and `admin` endpoints) and now to make a call by a different client you have to pass it inside of the methods above.
```
Comment (NoResponseMate) [inline UPGRADE-1.12.md]:
```suggestion
With this change, we reduced the number of clients from one per resource to only **two**, one for the `shop` section and one for the `admin`. To make a call for a specific resource you must pass it inside the methods shown above.
```

File: UPGRADE-1.12.md [hunk 4]
```diff
@@ -84,6 +84,115 @@ with test or remove these services with complier pass.
...
+You can see the actual difference in example below:
```
Comment (NoResponseMate) [inline UPGRADE-1.12.md]:
```suggestion
You can see the actual difference in the example below:
```

File: UPGRADE-1.12.md [hunk 5]
```diff
@@ -84,6 +84,115 @@ with test or remove these services with complier pass.
 9. Passing a `Gaufrette\Filesystem` to `Sylius\Component\Core\Uploader\ImageUploader` constructor is deprecated since
 Sylius 1.12 and will be prohibited in 2.0. Use `Sylius\Component\Core\Filesystem\Adapter\FlysystemFilesystemAdapter` instead.
 
+### Api clients refactor
+
+#### New way of defining the clients
+
+In `1.12` version we changed the way of defining API clients in our behat contexts.
```
Comment (NoResponseMate) [inline UPGRADE-1.12.md]:
```suggestion
In the `1.12` version we changed the way of defining API clients in our behat contexts.
```

File: UPGRADE-1.12.md [hunk 6]
```diff
@@ -84,6 +84,115 @@ with test or remove these services with complier pass.
...
+We expanded the  `Sylius/Behat/Client/ApiClientInterface` methods with extra parameter `$resource`.
```
Comment (NoResponseMate) [inline UPGRADE-1.12.md]:
```suggestion
We extended the following methods of `Sylius/Behat/Client/ApiClientInterface` with extra argument `$resource`.
```

File: UPGRADE-1.12.md [hunk 7]
```diff
@@ -84,6 +84,115 @@ with test or remove these services with complier pass.
 9. Passing a `Gaufrette\Filesystem` to `Sylius\Component\Core\Uploader\ImageUploader` constructor is deprecated since
 Sylius 1.12 and will be prohibited in 2.0. Use `Sylius\Component\Core\Filesystem\Adapter\FlysystemFilesystemAdapter` instead.
 
+### Api clients refactor
```
Comment (GSadee) [inline UPGRADE-1.12.md]:
```suggestion
### API clients refactor
```

Comment (GSadee) [inline UPGRADE-1.12.md]: And this chapter should be moved to `UPGRADE-API-1.12.md` file

File: UPGRADE-API-1.12.md [hunk 8]
```diff
@@ -117,3 +117,111 @@ Here is how the response looks like:
    has become nullable.
 
 1. The `Sylius\Bundle\ApiBundle\Assigner\OrderPromoCodeAssignerInterface` has been renamed to `Sylius\Bundle\ApiBundle\Assigner\OrderPromotionCodeAssignerInterface`.
+
+### API clients refactor
+
+#### New way of defining the clients
+
+In the `1.12` version we changed the way of defining API clients in our behat contexts.
+We expanded the  `Sylius/Behat/Client/ApiClientInterface` methods with extra parameter `$resource`.
```
Comment (GSadee) [inline UPGRADE-API-1.12.md]:
```suggestion
We extended the `Sylius/Behat/Client/ApiClientInterface` methods with extra parameter `$resource`.
```

File: UPGRADE-API-1.12.md [hunk 9]
```diff
@@ -117,3 +117,111 @@ Here is how the response looks like:
...
+With this change, we reduced the number of clients from one per resource to only **two**, one for the `shop` section and one for the `admin`. To make a call for a specific resource you must pass it inside the methods shown above.
+You can see the actual difference in the example below:
+
+  ```diff
+  -  $this->avatarImagesClient->buildUploadRequest();
+  +  $this->client->buildUploadRequest(Resources::AVATAR_IMAGES);
+  ```
+
+The `Sylius\Behat\Context\Api\Resources` class contains constants for all the defined resources.
+We also removed the `Sylius\Behat\ClientApiIriClient` alongside with `Sylius\Behat\ClientApiIriClientInterface` as it's no more used.
```
Comment (GSadee) [inline UPGRADE-API-1.12.md]:
```suggestion
We also removed the `Sylius\Behat\Client\ApiPlatformIriClient` alongside with `Sylius\Behat\Client\ApiIriClientInterface` as it's no more used.
```

File: UPGRADE-API-1.12.md [hunk 10]
```diff
@@ -117,3 +117,111 @@ Here is how the response looks like:
...
+
+#### Changes in contexts
+
+The constructors of the behat contexts changed with the improvement.
```
Comment (GSadee) [inline UPGRADE-API-1.12.md]:
```suggestion
The constructors of the behat contexts have changed in relation to the above improvements.
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Kamil! :1st_place_medal:

**Pull request merged**